differential-drive-bot/drivetrain.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+import wpilib.drive
+import wpimath.controller
+import wpimath.kinematics
+import wpimath.units
+
+import math
+
+
+class Drivetrain:
+    """Represents a differential drive style drivetrain."""
+
+    MAX_SPEED = 3.0  # 3 meters per second
+    MAX_ANGULAR_SPEED = math.pi  # 1/2 rotation per second
+
+    TRACK_WIDTH = 0.381 * 2  # meters
+    WHEEL_RADIUS = 0.0508  # meters
+    ENCODER_RESOLUTION = 4096  # counts per revolution
+
+    def __init__(self):
+        leftLeader = wpilib.PWMSparkMax(1)
+        leftFollower = wpilib.PWMSparkMax(2)
+        rightLeader = wpilib.PWMSparkMax(3)
+        rightFollower = wpilib.PWMSparkMax(4)
+
+        self.leftGroup = wpilib.MotorControllerGroup(leftLeader, leftFollower)
+        self.rightGroup = wpilib.MotorControllerGroup(rightLeader, rightFollower)
+
+        self.leftEncoder = wpilib.Encoder(0, 1)
+        self.rightEncoder = wpilib.Encoder(2, 3)
+
+        self.leftPIDController = wpimath.controller.PIDController(1.0, 0.0, 0.0)
+        self.rightPIDController = wpimath.controller.PIDController(1.0, 0.0, 0.0)
+
+        self.gyro = wpilib.AnalogGyro(0)
+
+        self.kinematics = wpimath.kinematics.DifferentialDriveKinematics(self.TRACK_WIDTH)
+        self.odometry = wpimath.kinematics.DifferentialDriveOdometry(
```

Comment:
This has to happen after the encoder is reset. See the Java example.

... which makes me think the C++ example is incorrect, actually. I'll look into that.

Reply:
https://github.com/wpilibsuite/allwpilib/issues/6013